example/amd_smi_list.cc

```diff
@@ -0,0 +1,169 @@
+        const char *err_str; \
+        std::cout << "AMDSMI call returned " << RET << " at line " \
+                    << __LINE__ << std::endl; \
+        amdsmi_status_code_to_string(RET, &err_str); \
+        std::cout << err_str << std::endl; \
+        return RET; \
+    } \
+}
+
+
+int main(int argc, char **argv) {
+    amdsmi_status_t ret;
+    uint32_t socket_count = 0;
+
+    // Initialize amdsmi for AMD CPUs
+    ret = amdsmi_init(AMDSMI_INIT_AMD_GPUS);
+
+    ret = amdsmi_get_socket_handles(&socket_count, nullptr);
+
+    // Allocate the memory for the sockets
+    std::vector<amdsmi_socket_handle> sockets(socket_count);
+
+    // Get the sockets of the system
+    ret = amdsmi_get_socket_handles(&socket_count, &sockets[0]);
+
+    std::cout << "Total Socket: " << socket_count << std::endl;
+
+    // For each socket, get cpus
+    for (uint32_t i = 0; i < socket_count; i++) {
+
+        // Get the device count available for the socket.
+        uint32_t device_count = 0;
+        uint32_t tmp_device_count = 0;
+        bool isGPU, isNIC, isSWITCH;
+        ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_AMD_GPU, nullptr, &tmp_device_count);
+        CHK_AMDSMI_RET(ret)
+
+        // Allocate the memory for the device handlers on the socket
+        
+        if (tmp_device_count == 0) { //it could be BRCM NIC device
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_NIC, nullptr, &tmp_device_count);
+          device_count = tmp_device_count;
+          isNIC = true;
+          CHK_AMDSMI_RET(ret)
+        } else {
+          device_count = tmp_device_count;
+          isGPU = true;
+        }
+
+        if (tmp_device_count == 0) {  // it could be BRCM SWITCH device
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH, nullptr,
+                                                     &tmp_device_count);
+          device_count = tmp_device_count;
+          isNIC = false;
+          isGPU = false;
+          isSWITCH = true;
+          CHK_AMDSMI_RET(ret)
+        }
+
+        // Allocate the memory for the device handlers on the socket
+        std::vector<amdsmi_processor_handle> processor_handles(device_count);
+
+        if (isNIC) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_NIC, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isGPU) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_AMD_GPU, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else if (isSWITCH) {
+          ret = amdsmi_get_processor_handles_by_type(sockets[i], AMDSMI_PROCESSOR_TYPE_BRCM_SWITCH, &processor_handles[0], &device_count);
+          CHK_AMDSMI_RET(ret)
+        } else {
+          printf("\n Unknown device as part of discovered list. Exiting! \n");
```

Comment:
There is a mix of `printf()` and `std::cout`. Let's stick to `std::cout` for consistency.